library/n2_protocol.hpp

```diff
@@ -13,36 +13,57 @@ struct data_in_file {
 	off_t local_offset = 0;
 	size_t fsize = 0;
 	int on_exit = 0;
+
+	void detach();
+	data_pointer read_and_detach();
 };
 
 struct data_place {
+	data_in_file in_file;
+	data_pointer in_memory;
+
 	enum place_type {
 		IN_MEMORY,
 		IN_FILE,
 	};
-
-	data_in_file in_file;
-	data_pointer in_memory;
-
 	place_type where() const;
 
-	static data_place from_file(const data_in_file &in_file);
-	static data_place from_memory(const data_pointer &in_memory);
+	void force_memory();
+
+	static data_place from_file(data_in_file in_file);
+	static data_place from_memory(data_pointer in_memory);
 };
 
 struct message {
 	dnet_cmd cmd;
 
-	virtual void make_owning() {}
 	virtual ~message() = default;
 };
 
+struct request_info {
+	std::unique_ptr<message> request;
+
+	// Ways to reply (list will be extended for streaming repliers)
+	std::function<void (std::shared_ptr<message>)> reply;
+	std::function<void (int)> reply_error;
+};
+
+struct response_info {
+	dnet_cmd cmd; // for log, for routing (in case of error-response we not always have message to extract cmd)
+	std::function<void ()> response_processor;
+};
+
+int send_response(struct dnet_net_state *st, struct dnet_cmd *cmd, std::function<void ()> response,
+                  struct dnet_access_context *context);
+
 }}} // namespace ioremap::elliptics::n2
 
 // All requests/responses must be inherited from n2_message;
 // common_request is visible from dnet_io_req. Since this struct
 // is visible from C code, it is declared out of namespace.
 struct n2_message : ioremap::elliptics::n2::message {};
+struct n2_request_info : public ioremap::elliptics::n2::request_info {};
```

Comment:
All these entities are superfluous. @shaitan will not approve it, how he did it in my PR.

Reply:
Let @shaitan to review it.